docs/proposals/20200506-conditions.md

```diff
@@ -0,0 +1,550 @@
+     - [Status field](#status-field)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+     - [Test Plan [optional]](#test-plan-optional)
+     - [Graduation Criteria [optional]](#graduation-criteria-optional)
+     - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+Condition: The state of an object with regard to its appearance, quality, or working order.
+
+## Summary
+
+In Cluster API a workload cluster is composed of a fair number of Kubernetes objects, and to
+understand the status of your Cluster you have to jump from one object to another. Or, in the
+worst case, you have to look at the logs from the Cluster API pods to get more details on why
+things aren’t working.
+
+This proposal introduces the concept of “conditions” as a tool designed to provide an “at a glance”
+view of the status of a cluster, providing immediate answers and fine-grained guidance when
+investigating issues. 
+
+## Motivation
+
+- Cluster API requires a large number of objects to define a workload cluster: Cluster,
+  InfrastructureCluster, KubeadmControlPlane, MachineDeployments, Machines, and so on.
+  Each object carries its own status. Operators need to describe each object individually to
+  determine an overall cluster health summary.
+- Objects can be tied to each other directly (via object references), or indirectly
+  (via owner references). These dependencies can often require a deep knowledge of the
+  project’s internals and its APIs to investigate issues or problems. 
+- Some objects, like MachineSets or MachineDeployments, are considered a collection of
+  other objects (e.g. MachineSets). Their status is effectively tied to the status of each
+  item in the collection. 
+
+### Goals
+
+- To define a data model for the Condition type.
+- To introduce the first set of conditions on Cluster API core objects (up to a minimal level of
+  detail required for an initial implementation).
+- To provide flexibility for providers to incrementally adopt this proposal.
+- To provide flexibility for providers to add, remove, or update their own set of conditions over time.
+- To provide utilities and libraries for a unified way to work with conditions.
+
+### Non-Goals/Future Work
+
+- To pre-define every possible condition.
+- To define the exact semantic for every condition introduced in the proposal (this should be done
+  during implementation).
+- To remove or deprecate `Phase`, `FailureMessage`, and `FailureReason` from Status in v0.3.
+- To remove or deprecate usage of `Events`.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+As a developer, as a user, as a tool built on top of Cluster API, I would like to have common types
+for defining conditions on different Cluster API or provider objects.
```

Comment:
Because all of them could benefit from if the same data type is used across all the Cluster API objects. 
BTW this is the same motivation for https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1623-standardize-conditions, so there is alignment with Kubernetes here

Reply:
That makes sense, maybe it's more a motivation than a user story, but it's fine.